Start removing `rustc_middle::hir::map::Map`

nnethercote opened the pull request:
`rustc_middle::hir::map::Map` is now just a low-value wrapper around `TyCtxt`. This PR starts removing it.

r? @cjgillot

nnethercote committed:
Rename `rustc_middle/src/hir/map/mod.rs` as `map.rs`.

There is no need for the extra subdirectory, and this makes the `map`
module consistent with its sibling modules `nested_filter` and `place`.

nnethercote committed:
Move some `Map` methods onto `TyCtxt`.

The end goal is to eliminate `Map` altogether.

I added a `hir_` prefix to all of them, that seemed simplest. The
exceptions are `module_items` which became `hir_module_free_items` because
there was already a `hir_module_items`, and `items` which became
`hir_free_items` for consistency with `hir_module_free_items`.

nnethercote committed:
Remove unused `Map::hir_node_by_def_id` method.

nnethercote committed:
Overhaul the `intravisit::Map` trait.

First of all, note that `Map` has three different relevant meanings.
- The `intravisit::Map` trait.
- The `map::Map` struct.
- The `NestedFilter::Map` associated type.

The `intravisit::Map` trait is impl'd twice.
- For `!`, where the methods are all unreachable.
- For `map::Map`, which gets HIR stuff from the `TyCtxt`.

As part of getting rid of `map::Map`, this commit changes `impl
intravisit::Map for map::Map` to `impl intravisit::Map for TyCtxt`. It's
fairly straightforward except various things are renamed, because the
existing names would no longer have made sense.

- `trait intravisit::Map` becomes `trait intravisit::HirTyCtxt`, so named
  because it gets some HIR stuff from a `TyCtxt`.
- `NestedFilter::Map` assoc type becomes `NestedFilter::MaybeTyCtxt`,
  because it's always `!` or `TyCtxt`.
- `Visitor::nested_visit_map` becomes `Visitor::maybe_tcx`.

I deliberately made the new trait and associated type names different to
avoid the old `type Map: Map` situation, which I found confusing. We now
have `type MaybeTyCtxt: HirTyCtxt`.

nnethercote committed:
Clarify `impl Map for !`.

All the methods are unreachable, so make that clearer.

nnethercote committed:
Avoid unnecessary use of the `Map` trait.

The `Map` trait is there for cases where `tcx` isn't available. This
isn't one of those cases, so it's simpler to just call through `tcx`
directly.

nnethercote commented:
There are a lot more methods to move from `Map` to `TyCtxt`, but I thought I would file this to see if people are happy with the general approach before sinking more time into it. It's good to see the changes to the `intravisit::Map` trait worked out.

Zalathar commented on compiler/rustc_hir/src/intravisit.rs:
```diff
@@ -106,45 +106,41 @@ impl<'a> FnKind<'a> {
 }
 
-/// An abstract representation of the HIR `rustc_middle::hir::map::Map`.
-pub trait Map<'hir> {
```
If I understand correctly, this whole abstraction only exists to work around the fact that `Map`/`TyCtxt` isn't visible from the `rustc_hir` crate, and in fact every implementation is either (`rustc_middle::hir::Map` → `TyCtxt`) or `!`.

So it might be clearer to have the naming of this trait, its methods, and `nested_visit_map` reflect the fact that this is really just a `TyCtxt` in practice.

nnethercote replied:
That's exactly what I tried to do with the `HirCtxt`/`Cx` names. I'm happy to hear alternative suggestions.

Zalathar replied:
Ah, to be clear what I had in mind with my suggestion was a naming scheme like:

- `trait HirTyCtxt`
  - Every trait method would have exactly the same name as its real `TyCtxt` equivalent
- In the HIR visitor traits:
  - Associated `type TyCtxt`
  - `fn nested_visit_tcx()` (or even just `fn tcx()`)

The idea would be to signal to the reader “you can assume that this is really just a TyCtxt` under the hood”.

nnethercote replied:
Ok, I did a lot of that, though I went with `MaybeTyCtxt` for the associated type and `maybe_tcx` for the HIR visitor method.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13108024162/job/36566016896) [(plain)](https://github.com/rust-lang/rust/commit/14d4561d58d12a3b2702eeae960d2f1ef11e4602/checks/36566016896/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
  --> compiler/rustc_passes/src/hir_id_validator.rs:64:36
   |
64 |             let pretty_owner = hir.def_path(owner.def_id).to_string_no_crate_verbose();
   |                                    ^^^^^^^^ method not found in `Map<'_>`
error[E0308]: mismatched types
   --> compiler/rustc_passes/src/hir_id_validator.rs:109:9
    |
    |
108 |     fn nested_visit_cx(&mut self) -> Self::Cx {
    |                                      -------- expected `TyCtxt<'hir>` because of return type
109 |         self.tcx.hir()
    |         ^^^^^^^^^^^^^^ expected `TyCtxt<'_>`, found `Map<'_>`
error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_passes/src/hir_id_validator.rs:141:36
    |
    |
141 |                     self.tcx.hir().def_path(hir_id.owner.def_id).to_string_no_crate_verbose(),
    |                     --------       ^^^^^^^^ method not found in `Map<'_>`
    |                     method `def_path` is available on `TyCtxt<'hir>`

error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_passes/src/hir_id_validator.rs:142:36
   --> compiler/rustc_passes/src/hir_id_validator.rs:142:36
    |
142 |                     self.tcx.hir().def_path(owner.def_id).to_string_no_crate_verbose()
    |                     --------       ^^^^^^^^ method not found in `Map<'_>`
    |                     method `def_path` is available on `TyCtxt<'hir>`

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13129372091/job/36631493606) [(plain)](https://github.com/rust-lang/rust/commit/90a871b06fd0de23fb120816a81afa1b07eba3da/checks/36631493606/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.947 Building wheels for collected packages: reuse
#16 2.948   Building wheel for reuse (pyproject.toml): started
#16 3.164   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.165   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=be6760d5849de4a58bbe52b85ca57a55f2b32b518b17029a5ad2e530db0d4303
#16 3.165   Stored in directory: /tmp/pip-ephem-wheel-cache-jmg4bsdy/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.168 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.567 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.567 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.7s
---
    Checking rustc_hir_typeck v0.0.0 (/checkout/compiler/rustc_hir_typeck)
error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
  --> compiler/rustc_passes/src/hir_id_validator.rs:64:36
   |
64 |             let pretty_owner = hir.def_path(owner.def_id).to_string_no_crate_verbose();
   |                                    ^^^^^^^^ method not found in `Map<'_>`
error[E0308]: mismatched types
   --> compiler/rustc_passes/src/hir_id_validator.rs:109:9
    |
    |
108 |     fn maybe_tcx(&mut self) -> Self::MaybeTyCtxt {
    |                                ----------------- expected `TyCtxt<'hir>` because of return type
109 |         self.tcx.hir()
    |         ^^^^^^^^^^^^^^ expected `TyCtxt<'_>`, found `Map<'_>`
error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_passes/src/hir_id_validator.rs:141:36
    |
    |
141 |                     self.tcx.hir().def_path(hir_id.owner.def_id).to_string_no_crate_verbose(),
    |                     --------       ^^^^^^^^ method not found in `Map<'_>`
    |                     method `def_path` is available on `TyCtxt<'hir>`

error[E0599]: no method named `def_path` found for struct `rustc_middle::hir::map::Map` in the current scope
   --> compiler/rustc_passes/src/hir_id_validator.rs:142:36
   --> compiler/rustc_passes/src/hir_id_validator.rs:142:36
    |
142 |                     self.tcx.hir().def_path(owner.def_id).to_string_no_crate_verbose()
    |                     --------       ^^^^^^^^ method not found in `Map<'_>`
    |                     method `def_path` is available on `TyCtxt<'hir>`

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13130827641/job/36635473123) [(plain)](https://github.com/rust-lang/rust/commit/2ed2d93c3a3de9533119093ecdba24a1bdd67248/checks/36635473123/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.2s done
#22 DONE 36.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui-fulldeps/obtain-borrowck.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui-fulldeps/obtain-borrowck.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/obtain-borrowck/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021"
--- stderr -------------------------------
error[E0599]: no method named `trait_item` found for struct `rustc_middle::hir::map::Map` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/obtain-borrowck.rs:82:38
   |
   |
LL |                 let trait_item = hir.trait_item(id);
   |
help: there is a method `expect_trait_item` with a similar name
   |
   |
LL |                 let trait_item = hir.expect_trait_item(id);

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0599`.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #136549) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13150221828/job/36696242963) [(plain)](https://github.com/rust-lang/rust/commit/f5383a166d457ce19f0e7be41b932dc705e40192/checks/36696242963/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_infer v0.0.0 (/checkout/compiler/rustc_infer)
error: unresolved link to `Map`
   --> compiler/rustc_hir/src/intravisit.rs:241:11
    |
241 |     /// [`Map`]; see `maybe_tcx` for advice.
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `rustc_hir`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Wed Feb  5 04:54:35 UTC 2025
  network time: Wed, 05 Feb 2025 04:54:35 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13150541827/job/36697093192) [(plain)](https://github.com/rust-lang/rust/commit/daae919e2db8e387b312e7d71697e3cf5be0341b/checks/36697093192/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
error: unresolved link to `TyCtxt`
   --> compiler/rustc_hir/src/intravisit.rs:241:11
    |
241 |     /// [`TyCtxt`]; see `maybe_tcx` for advice.
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `rustc_hir`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Wed Feb  5 05:28:54 UTC 2025
  network time: Wed, 05 Feb 2025 05:28:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

nnethercote commented:
Note: [this comment](https://github.com/rust-lang/rust/pull/118256#issuecomment-1826442834) is the one that identified that `Map` is now just a wrapper around `TyCtxt`.

bors commented:
:umbrella: The latest upstream changes (presumably #136613) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased. This PR is conflict-prone.

bors commented:
:umbrella: The latest upstream changes (presumably #136878) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
The conflicts are trivial and won't affect review, so I will wait before fixing them.

nnethercote commented:
It's been two weeks, let's try a different reviewer. I have rebased to fix the conflicts.

r? Zalathar

cjgillot commented on compiler/rustc_middle/src/hir/map.rs:
```diff
@@ -164,76 +164,80 @@ impl<'tcx> TyCtxt<'tcx> {
-        self.tcx.hir_crate(())
+    pub fn hir_krate(self) -> &'tcx Crate<'tcx> {
+        self.hir_crate(())
     }
```
Should we just remove this one?

nnethercote replied:
ok

cjgillot commented on compiler/rustc_middle/src/hir/map.rs:
```diff
@@ -164,76 +164,80 @@ impl<'tcx> TyCtxt<'tcx> {
     #[inline]
-    pub fn items(self) -> impl Iterator<Item = ItemId> + 'hir {
-        self.tcx.hir_crate_items(()).free_items.iter().copied()
+    pub fn hir_items(self) -> impl Iterator<Item = ItemId> + 'tcx {
```
`hir_free_items` would be more descriptive and consistent with `hir_module_free_items`

cjgillot reviewed (commented):
Thanks for cleaning this up. 
2 nits and r=me

nnethercote committed:
Remove `TyCtxt::hir_krate`.

It's a trivial wrapper around the `hir_crate` query with a small number
of uses.

nnethercote commented:
I fixed the nits.

@bors r=cjgillot

bors commented:
:pushpin: Commit f666361caa963c3f41934d5b431af24d8894d159 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot f666361caa963c3f41934d5b431af24d8894d159 -->
<!-- homu: {"type":"Approved","sha":"f666361caa963c3f41934d5b431af24d8894d159","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
